Milestone 2.2: Upgrade explorations that use the old math interaction.

iamprayush opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

~~1. This PR fixes or fixes part of #[fill_in_number_here].~~
2. This PR does the following:
Adds a state migration that upgrades all explorations that use the old math interaction.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

iamprayush committed:
Added upgradation job. [skip ci]

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations [skip-ci]

iamprayush committed:
Fixed linter errors. [skip ci]

seanlip commented on core/jobs_registry.py:
```diff
@@ -60,6 +60,7 @@
     exp_jobs_one_off.InteractionCustomizationArgsValidationJob,
     exp_jobs_one_off.ItemSelectionInteractionOneOffJob,
     exp_jobs_one_off.MathExpressionValidationOneOffJob,
+    exp_jobs_one_off.MathExpressionUpgradationOneOffJob,
```
"Upgradation" is not a word. I think you want "Upgrade".

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -77,6 +78,82 @@
+    trig_fns = ['sin', 'cos', 'tan', 'csc', 'sec', 'cot']
+
+    # Shifting powers in trig functions to the end.
+    # For eg. 'sin^2(x)' -> 'sin(x)^2'.
```
What will this look like, visually -- could you show a screenshot?

iamprayush replied:
Oh, I just realized this should be `(sin(x))^2` instead. I've corrected it. Also, in the editor, this is how it would look like:
![mathEditor](https://user-images.githubusercontent.com/35144226/87132119-0c543080-c2b3-11ea-8d37-e80d30eea569.png)

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -77,6 +78,82 @@
+        math_expression = re.sub(
+            r'%s(?!\()(.)' % trig_fn, r'%s(\1)' % trig_fn, math_expression)
+
+    # The pylatexenc lib outputs the unicode values of
```
What are you going to do to ensure that your dict is comprehensive and there are no remaining unicode chars that have not been handled?

iamprayush replied:
I have added the unicode characters that we support in the mapping dict and they will be replaced with their string equivalents. For the remaining ones, since they are not supported, they won't be replaced and will render the input as invalid, which is the expected behavior.

seanlip replied:
Where does this "render the input as invalid" happen in the code?

iamprayush replied:
While tokenizing the expression, the following regex would skip the invalid character and the check following these lines would raise the `Invalid token` exception.
https://github.com/oppia/oppia/blob/d1544d8f74b9e399e3b5aefe940b07cab1e7d974/core/domain/expression_parser.py#L132-L151

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -77,6 +78,82 @@
+VALID_MATH_INPUTS_YIELD_LIMIT = 200
+
+
+def clean_math_expression(math_expression):
```
Probably should make this private. (In general make things private where possible.)

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                            rule_input = clean_math_expression(rule_input)
 
                             validity = 'Invalid'
                             if is_valid_math_expression(rule_input):
```
Just a thought -- would it be better to have is_valid_algebraic_expression() and is_valid_numeric_expression() instead? When I read is_valid_math_expression() it doesn't automatically suggest an algebraic expression.

iamprayush replied:
That was the original plan, but the thing is since both the functions have a lot of overlapping code, I chose to refactor them into a single function with a parameter to specify the type of expression. I understand that seeing `is_valid_math_expression` doesn't imply "algebraicness" which is why I've removed the default parameter so that we need to explicitly mention upon each call whether the expression is algebraic/numeric.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                            elif is_valid_math_expression(
+                                    rule_input, algebraic=False):
+                                validity = 'Valid Numeric Expression'
                             elif is_valid_math_equation(rule_input):
```
Same thing (possibly) if this is checking for algebraic-ness too.

iamprayush replied:
Addressed above.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
-                                validity = 'Valid Equation'
+                                validity = 'Valid Math Equation'
+
+                            if validity != 'Invalid':
```
If you use a string more than once, always make it a named constant.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                                    'The exploration with ID: %s and state '
+                                    'name: %s contains inputs that correspond '
+                                    'to two different types. Please resolve '
+                                    'this before running the upgradation '
```
"upgradation" is not a word. Perhaps "upgrade".

Ditto elsewhere.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                                yield ('ERROR', (
+                                    'The exploration with ID: %s and state '
+                                    'name: %s contains inputs that correspond '
+                                    'to two different types. Please resolve '
```
Perhaps you should say which types.

Also, > 1 does not imply 2.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                        new_answer_groups.append(new_answer_group.to_dict())
+
+                    exp_services.update_exploration(
+                        'admin', item.id, [exp_domain.ExplorationChange({
```
Probably OppiaMigrationBot, not admin.

Also, again, never hardcode strings that are used more than once in the codebase, please. Use named constants. There are many other examples of this in the PR -- please fix this.

iamprayush replied:
Understood, done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+            for state_name, state in exploration.states.items():
+                if state.interaction.id == 'MathExpressionInput':
+                    new_answer_groups = []
+                    new_interaction_id = ''
```
You should set this to a valid value or at least validate at the end that it is not ''.

iamprayush replied:
Done.

seanlip replied:
Not done. I think there might be some confusion about what I meant by "validate". What I'm trying to say is that you should fail visibly somehow if this is still '' at the end, because presumably that is a non-expected case. Currently that case seems to be just swept under the rug.

Also why is this ''? Shouldn't it be None?

iamprayush replied:
> Not done. I think there might be some confusion about what I meant by "validate". What I'm trying to say is that you should fail visibly somehow if this is still '' at the end, because presumably that is a non-expected case. Currently that case seems to be just swept under the rug.

Understood, I've added an explicit ERROR message for that case.

>  Also why is this ''? Shouldn't it be None?

I've changed the code so that it is consistent in both the jobs and easier to follow, which is why this is not required.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                                new_interaction_id = 'NumericExpressionInput'
+                            elif is_valid_math_equation(rule_input):
+                                validity = 'Valid Math Equation'
+                                new_interaction_id = 'MathEquationInput'
```
You still need to raise/yield/etc. if an error is discovered, and handle that case safely, providing enough information to the admin such that they can take manual measures if needed or re-run the job again.

Do not leave anything to chance in a migration. Mistakes happen, so program defensively and do not assume anything.

iamprayush replied:
Done.

seanlip replied:
Not done. There needs to be an "else" clause.

iamprayush replied:
I've handled this now. If any of the inputs are categorized as Invalid, an error will be generated and that state won't be upgraded.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,68 +256,133 @@ def map(item):
+                        'admin', item.id, [exp_domain.ExplorationChange({
+                            'cmd': 'edit_state_property',
+                            'state_name': state_name,
+                            'property_name': 'widget_id',
```
No changes to customization args?

Also, when an interaction is changed normally through the UI, is this the list of changes that typically results? Please explain in detail what steps you took to verify this.

iamprayush replied:
> No changes to customization args?

There are no customization args in the old interaction or in the new one so that won't require changes. 

>  Also, when an interaction is changed normally through the UI, is this the list of changes that typically results? Please explain in detail what steps you took to verify this.

Yes. So, I created two explorations, one with the old interaction and the other with the new one. I then downloaded the explorations in the yaml format and I saw that these are the attributes that are different in both interactions. Also, I have tested this upgrade job locally and it  works as expected.

seanlip commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -939,6 +1057,176 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        self.albert_id = self.get_user_id_from_email(self.ALBERT_EMAIL)
+        self.process_and_flush_pending_tasks()
+
+    def test_explorations_are_properly_upgraded(self):
```
Write more tests. Cover all corner and unexpected cases. Ensure that the job is idempotent if something goes wrong midway.

iamprayush replied:
Done.

seanlip replied:
Where is the test for idempotence? Simulate the case where a migration goes wrong midway, and then the migration is re-run. Verify that the result of this is the same as the migration running fully the first time.

iamprayush replied:
I've modified the `test_explorations_with_invalid_inputs_are_not_upgraded` method to replicate the behavio that you've mentioned. Does that look alright?

seanlip replied:
Yes, thanks. I've added a comment but in general it looks good!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Resolved comments.

iamprayush committed:
Fixed linter errors [skip ci]

aks681 commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -77,6 +78,86 @@
 AUDIO_DURATION_SECS_MIN_STATE_SCHEMA_VERSION = 31
+# This threshold puts a cap on the number of valid inputs, i.e.,
+# expressions/equations that can be yielded by the math expression one-off jobs.
+VALID_MATH_INPUTS_YIELD_LIMIT = 200
```
Why  is this limit present?

iamprayush replied:
This limits the number of valid inputs of each type that are yielded by the job. We're displaying the valid outputs only for confirming that they are correct in general. We don't need to closely inspect each of the valid inputs, which is why we aren't displaying all of them. I've updated the comment here to make things clearer.

aks681 commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,69 +260,148 @@ def map(item):
+
+
+class MathExpressionUpgradeOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type (expression/equation) of the inputs present
+    in the exploration.
```
It also upgrades the exploration as well, right?

iamprayush replied:
Ah yes, I've updated the comment.

aks681 commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,69 +260,148 @@ def map(item):
 
+                        new_answer_groups.append(new_answer_group.to_dict())
+
+                    if new_interaction_id != '':
```
if new_interaction_id does become '', then that means, it didn't fall into any one case above, right? Handle that as well, by yielding another message?

iamprayush replied:
I've changed the code so that it is consistent in both the jobs and easier to follow, which is why this is not required.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -366,24 +375,33 @@ def map(item):
+                                    'state_name': state_name,
+                                    'property_name': 'answer_groups',
+                                    'new_value': new_answer_groups
+                                })], 'Upgraded interaction to %s.' % (
```
>> Also, when an interaction is changed normally through the UI, is this the list of changes that typically results? Please explain in detail what steps you took to verify this.

> Yes. So, I created two explorations, one with the old interaction and the other with the new one. I then downloaded the explorations in the yaml format and I saw that these are the attributes that are different in both interactions. Also, I have tested this upgrade job locally and it works as expected.

This is not a sufficient test. Please find a way to introspect the actual list of changes that is created when you do similar operations through the editor UI. Verify that that list of changes matches the changelist that you have here, and add that evidence to this comment.

iamprayush replied:
The thing is, it is not possible to change the type of interaction through the editor UI. The only way to replicate this behavior through the UI is to delete the state and create a new one with a different interaction. I did that and the changes are as expected.
![changelist](https://user-images.githubusercontent.com/35144226/87243890-5249f980-c457-11ea-9b7f-a5cd1e4396d2.gif)

aks681 replied:
What you are doing seems like the right way to achieve 'change interaction' in the frontend. Though, can you maybe log what is present in `UndoRedoService.getCommittableChangeList` or the change list received by the controller  in the backend, when you do 'Delete Interaction -> New Interaction -> Change rules in answer groups'? Since, that is what you are doing, right?

seanlip replied:
+1 -- what @aks681 explains is exactly the evidence I was looking for. Please could you provide it?

iamprayush replied:
@aks681 @seanlip Here are the logs that show the change list when I do the process of creating a new interaction with math expression input -> deleting the state -> creating a new state with Algebraic Expression Input with the same rule input.

![1](https://user-images.githubusercontent.com/35144226/87301799-cfe03900-c52d-11ea-9509-7124feb98377.png)
![2](https://user-images.githubusercontent.com/35144226/87301794-ceaf0c00-c52d-11ea-9f6f-75a5e9ef2a36.png)

Note that the migration job doesn't need to do this (deleting the state and creating a new one), the migration job can directly change the interaction ID which is why the exact changelist in the job is different but logically, they are both the same.

seanlip replied:
You keep talking about "deleting the state". Don't you mean "deleting the interaction"?

Also, please replicate exactly this changelist in your PR. Mimic what would happen in the frontend if you were making the changes manually. If we ever need to do any investigation retroactively in the future, then having everything follow a similar format would be helpful. It would make it easier to assume that the data we have is in a certain format.

Please fix this along the above lines and provide proof that it is done correctly (e.g. for adding the rules etc.). It is important that the backend commit cmd list match exactly whatever would have been done in the frontend had the same commands been performed manually.

iamprayush replied:
> You keep talking about "deleting the state". Don't you mean "deleting the interaction"?

No, I do mean "deleting the state". Each interaction has multiple states and each state can have a single interaction with multiple answer groups. So, this job checks for all inputs in all answer groups, and only if all of the inputs are valid and correspond to a single interaction, do we upgrade that state. The point is that states are independent from each other so invalidity of one state should not restrict the upgrade of the other just because they happen to be in the same exploration. So, in a way, I think "deleting the interaction" is the same is "deleting the state" since each state can have exactly one interaction.

> Also, please replicate exactly this changelist in your PR. Mimic what would happen in the frontend if you were making the changes manually. If we ever need to do any investigation retroactively in the future, then having everything follow a similar format would be helpful. It would make it easier to assume that the data we have is in a certain format.

Done. Now, the upgrade job will delete the state and replace it with a copy that has the changes -- just like it would happen in the frontend. I have simply updated the changelist that we update the interaction to match with the one that is generated when the operation is performed from the frontend as shown in the screenshot above.

iamprayush replied:
@seanlip, I actually meant "Each _exploration_ (not interaction) has multiple states...". 
And yes, I do mean deleting the interaction and not the state. I was confused between states and interactions.

/cc @aks681

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -81,9 +81,14 @@
 # This threshold puts a cap on the number of valid inputs, i.e.,
 # expressions/equations that can be yielded by the math expression one-off jobs.
 VALID_MATH_INPUTS_YIELD_LIMIT = 200
+_INVALID_EXPR = 'Invalid'
```
Please try to avoid using abbreviations when naming things.

In addition, each of these should be prefixed by a "type". For example we might have _STATUS_OPEN, _STATUS_CLOSED, _STATUS_IGNORED, etc. for feedback threads. Do something similar here so that it's clear that all these belong to the same "conceptual group". Also do this in general when naming groups of constants.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -128,11 +133,11 @@ def clean_math_expression(math_expression):
 
     # Shifting powers in trig functions to the end.
-    # For eg. 'sin^2(x)' -> 'sin(x)^2'.
+    # For eg. 'sin^2(x)' -> '(sin(x))^2'.
```
> Oh, I just realized this should be `(sin(x))^2` instead. I've corrected it. Also, in the editor, this is how it would look like:
![mathEditor](https://user-images.githubusercontent.com/35144226/87132119-0c543080-c2b3-11ea-8d37-e80d30eea569.png)

OK, thanks. The more usual way to type/display this is:

![Screenshot from 2020-07-10 16-48-48](https://user-images.githubusercontent.com/10575562/87211201-3ca8c880-c2cd-11ea-9960-482e19f1af07.png)

How would a user do that?

iamprayush replied:
Unfortunately, it is not possible to type/display `sin^2x` like this in the guppy editor.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,69 +260,148 @@ def map(item):
+                            rule_input = _clean_math_expression(rule_input)
+
+                            validity = _INVALID_EXPR
+                            if is_valid_math_expression(
```
> That was the original plan, but the thing is since both the functions have a lot of overlapping code, I chose to refactor them into a single function with a parameter to specify the type of expression. I understand that seeing is_valid_math_expression doesn't imply "algebraicness" which is why I've removed the default parameter so that we need to explicitly mention upon each call whether the expression is algebraic/numeric.

Oh, then this is what you do. Still have is_valid_algebraic_math_expression() etc. as your "public API". Behind the scenes, these functions call a private _is_valid_math_expression(True/False).

That way, you still reuse code while keeping this clear to clients.

iamprayush replied:
Ah, that makes sense. I've changed it now, thanks!

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -216,69 +260,148 @@ def map(item):
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
+        committer_id = 'OppiaMigrationBot'
```
No hardcoding. I've said this already.

This is already defined in the codebase as an existing named constant. Use that constant.

iamprayush replied:
Done.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Resolved review comments.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -77,6 +78,89 @@
+# This threshold puts a cap on the number of valid inputs, i.e.,
+# expressions/equations that can be yielded by the math expression one-off jobs.
+# The reason for limiting the number of valid inputs yielded by the jobs is that
+# we don't need to closesly inspect all of the valid inputs, they are displayed
```
Spelling: closely

iamprayush replied:
Fixed.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -77,6 +78,89 @@
+    # The pylatexenc lib outputs the unicode values of
+    # special characters like sqrt and pi, which is why
+    # they need to be replaced with their corresponding
+    # text values before performing validation.
```
Add: Other unicode characters will be left in the string as-is, and will be rejected by the expression parser.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+                    if _TYPE_INVALID_EXPRESSION in types_of_inputs:
+                        yield ('ERROR', (
+                            'There are some invalid inputs that need to be '
+                            'resolved before running the upgrade job.'))
```
Which ones? This message doesn't help at all with debugging.

iamprayush replied:
The invalid inputs are displayed in the output with the key 'Invalid'. I've added this in the error message as well to avoid confusion.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+                                type_of_input = _TYPE_VALID_MATH_EQUATION
+
+                            types_of_inputs.add(type_of_input)
+
```
At this point, shouldn't you be raising and breaking if the expression has invalid type?

You should not be doing any edits at all to an exploration if any "badness" is discovered.

iamprayush replied:
We don't wanna break upon finding invalid inputs, because we need to output them so that they can be resolved. And if any of the types_of_inputs is found invalid, we don't modify that state at all. This is checked after the loop on the answer groups is finished.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+
+                    if _TYPE_INVALID_EXPRESSION in types_of_inputs:
+                        yield ('ERROR', (
+                            'There are some invalid inputs that need to be '
```
This is not a helpful message. It does not help with debugging.

iamprayush replied:
The invalid inputs are displayed in the output with the key 'Invalid'. I've added this in the error message as well to avoid confusion.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
-            MathExpressionValidationOneOffJob.VALID_MATH_INPUTS_YIELD_LIMIT)
-        if key.startswith('Valid'):
-            yield (key, values[:valid_inputs_limit])
+        if key.endswith('Input'):
```
Be more specific here. If you're trying to say "if key is one of _TYPE_..." then say that. Otherwise it's hard for the reader of this to know what your intention is, and "Input" itself doesn't really convey much meaning compared to "Valid"/"Invalid".

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -939,6 +1058,295 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+
+        exploration = exp_fetchers.get_exploration_by_id('exp_id0')
+
+        # The exploration contains an invalid input and must therefore, not be
```
Validate also that no changes at all have been made to the exploration.

iamprayush replied:
I've now checked if the version remains the same or not. That should suffice, right?

seanlip replied:
Mm, sort of. It's just about acceptable, but it would be better to have a more thorough check of the content of the exploration itself since the version number is one step removed.

iamprayush replied:
Ok, I've updated the tests to check for the entire exploration content, instead of just the version.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Addressed review comments. [skip ci]

seanlip reviewed (approved):
Thanks @iamprayush! No further comments besides [this one](https://github.com/oppia/oppia/pull/9863#discussion_r454099291).

seanlip reviewed (changes requested):
Wait -- I can't approve this yet. Will need to run the backup job first (after the usual form is filled out).

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Made tests more thorough.

iamprayush commented:
> Wait -- I can't approve this yet. Will need to run the backup job first (after the usual form is filled out).

The form said that it should be filled after the PR has been approved and the tests have passed, which is why I haven't filled it yet. Also note, that https://github.com/oppia/oppia/pull/9822 will have to be merged before this one since this requires the numeric interaction in the codebase to run successfully.

seanlip commented:
Ah sorry I wasn't clear, my bad. Yup, I know that it's not time to fill in the form yet -- just wanted to explain why I had to rescind my approval and what I needed to wait on. Thanks!

aks681 reviewed (approved):
Lgtm!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed backend tests.

iamprayush committed:
Fixed backend tests.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed linter errors.

iamprayush commented:
Hey @seanlip, the backend tests have passed since the numeric expression's PR got merged. So now, this job is ready to be run. I've filled out the form, PTAL, thanks!

seanlip reviewed (changes requested):
@iamprayush, sorry, I missed replying to an earlier comment. PTAL.

Also, I did not get your form response. Could you try resubmitting? Maybe paste the link of the form you submitted to, for reference.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Updating changelist to match with the one that is generated in the frontend.

seanlip commented:
Please talk to Akshay. "Deleting the state" is still a misconception here
and your explanation has a bunch of errors. I suggest clearing that up with
him.

Also, make sure when you are testing in the frontend to try it with a case
that already has answer groups and rules. See if existing rules are kept
when you change the interaction.

iamprayush commented:
> Please talk to Akshay. "Deleting the state" is still a misconception here and your explanation has a bunch of errors. I suggest clearing that up with him.

Will do. I'll update the thread after.

> Also, make sure when you are testing in the frontend to try it with a case that already has answer groups and rules. See if existing rules are kept when you change the interaction.

Okay, so these are the scenarios I've tested:
1. One state, valid input. (Migrates as expected)
2. One state, invalid input. (Produces error, no change in data)
3. Two states, one valid another invalid. (The valid one gets upgraded, and the invalid one gets logged and no change is done to it).
4. Two states, both valid but with different inputs (one numeric another algebraic), (Both get upgraded to the corresponding new interactions).
5. Two states, one valid math expression input, and another random interaction. (Valid one gets upgraded as expected and the other interaction is unchanged).

Please let me know if I've missed something, thanks!

Also, I've filled out the job form again.

seanlip commented:
@iamprayush, you need to actually carefully read and follow the instructions in the job form if you want your job to be run. Your doc is not in the correct format and it does not have sufficient information.

Please update the doc as needed and then ping this thread again.

iamprayush commented:
@seanlip Sorry about that, I didn't realize that the format had changed. I have updated the [doc](https://docs.google.com/document/d/1QomDD3iOWgbTRdlLWGeSYxcVSbG-7B_7ftbQOLXL2T4/edit?usp=sharing), PTAL, thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Added conditions for resolving errors.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

seanlip commented:
As mentioned in email -- please convert the upgrade to a regular schema upgrade.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Added state migration functions.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed minor errors.

iamprayush committed:
temp commit

iamprayush committed:
Fixed minor error.

iamprayush committed:
Fixed minor error.

iamprayush committed:
Added tests.

iamprayush committed:
temp commit

iamprayush committed:
temp commit

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Changed draft upgrade function,

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Added draft upgrade tests.

iamprayush committed:
Minor refactoring.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed tests.

iamprayush committed:
Fixed linter errors.

iamprayush commented:
> As mentioned in email -- please convert the upgrade to a regular schema upgrade.

Done. Just to confirm, I have tested this locally on various scenarios and the behavior is as expected.

seanlip commented:
Please list the scenarios.

iamprayush commented:
> Please list the scenarios.

1. One state with one answer group and one rule input of each type (Algebraic expression, numeric expression and math equation) -> migrate to their respective new interaction types.

2. One state with one answer group with multiple rule inputs: one equation, and some expressions. -> migrated to equation type and expression rule types are deleted.

3. One state with two answer groups, one with one rule input of type equation, and the other with one rule input of type expression. -> state gets migrated to equation type and the 2nd answer group is entirely deleted since it does not contain any equations.

4. Performed the same scenarios with algebraic and numeric expression and the numeric expression inputs were deleted leaving only the algebraic types.

5. One state with only numeric inputs -> gets migrated to the numeric type.

6. Exploration with unsaved changes (some containing math interaction inputs) -> Remains untouched.

7. Also tested with saved explorations that are not yet published and they got migrated as expected.

Please let me know if I've missed something, I will test that as well, thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed backend test failures.

iamprayush commented on core/domain/user_jobs_one_off.py:
```diff
@@ -18,13 +18,9 @@
 from __future__ import unicode_literals  # pylint: disable=import-only-modules
 
 import ast
-import logging
```
I've removed this because this job was temporary and was failing when I tried upgrading the version from 34 to 35 in the tests for this job. This is in order to mirror the changes made by @donosco98 as seen in [this commit](https://github.com/oppia/oppia/pull/9985/commits/89342471240d7598c2c1654b68fc7d37df14fca0).

DubeySandeep reviewed (approved):
LGTM for the codeowner files!

iamprayush committed:
Partially fixed backend coverage [skip ci]

aks681 commented on core/domain/exp_domain.py:
```diff
@@ -2529,6 +2611,99 @@ def _convert_states_v33_dict_to_v34_dict(cls, states_dict):
+                            len(answer_group['rule_specs']) != 0)]
+
+                    state_dict['interaction']['id'] = new_interaction_id
+                    state_dict['interaction']['answer_groups'] = (
```
Wouldn't this affect solutions also?

iamprayush replied:
Yeah, I've added that as well now.

aks681 commented on core/domain/draft_upgrade_services.py:
```diff
@@ -164,6 +164,23 @@ def extract_html_from_draft_change_list(draft_change_list):
+        Returns:
+            list(ExplorationChange). The converted draft_change_list.
+        """
+        return draft_change_list
```
I thought the delete would be done here? Or is  that after the audit job is run?

iamprayush committed:
Fixed backend coverage and resolved review comments.

iamprayush committed:
Fixed linter errors.

aks681 commented on core/domain/draft_upgrade_services.py:
```diff
@@ -164,6 +164,31 @@ def extract_html_from_draft_change_list(draft_change_list):
+        """
+        for change in draft_change_list:
+            # We don't want to migrate any changes that involve the
+                # MathExpressionInput interaction.
```
Wrong indentation.

iamprayush replied:
Fixed.

aks681 commented on core/domain/draft_upgrade_services.py:
```diff
@@ -164,6 +164,31 @@ def extract_html_from_draft_change_list(draft_change_list):
+            if (change.property_name ==
+                    exp_domain.STATE_PROPERTY_INTERACTION_ID and (
+                        change.new_value == 'MathExpressionInput')):
+                raise Exception('Conversion cannot be completed.')
```
Where is the try catch where  the conversion function is called that catches this error and returns None?

iamprayush replied:
Added.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Resolved review comments.

seanlip commented:
@iamprayush Lint checks are failing. How did this happen?

iamprayush commented:
> @iamprayush Lint checks are failing. How did this happen?

@seanlip It failed while Installing third-party dependencies. Should I restart the tests? The lighthouse tests are also failing.

seanlip commented on core/domain/draft_upgrade_services.py:
```diff
@@ -164,6 +167,39 @@ def extract_html_from_draft_change_list(draft_change_list):
+
+        Returns:
+            list(ExplorationChange). The converted draft_change_list.
+        """
```
Missing Raises docstring.

iamprayush replied:
Done.

seanlip reviewed (commented):
Yup, do whatever you need to do in order to get the CI tests to pass.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed docstring.

iamprayush committed:
Fixed docstring.

seanlip commented:
@iamprayush Make sure to fill out the form if you want this job to be run.

iamprayush commented:
> @iamprayush Make sure to fill out the form if you want this job to be run.

Yeah, I have the form ready and I've also updated the instructions doc, I was trying to make sure that the CI checks pass before submitting it. I have fixed everything now so as soon as the backend tests pass, I'll submit the form.

seanlip commented:
I agree that that is the normal procedure. But if you want the job to run in the next 30 minutes rather than in the next 12 hours, and you do not expect the form submission contents to change, then I suggest submitting it now.

iamprayush commented:
> I agree that that is the normal procedure. But if you want the job to run in the next 30 minutes rather than in the next 12 hours, and you do not expect the form submission contents to change, then I suggest submitting it now.

I've submitted the form now.

seanlip commented:
Just an update -- I did a spot-check on some migrated explorations and found several issues:

(1) When I type stuff like 5x^2 + ..., a red error message appears and disappears at the bottom while I'm typing. This is very disconcerting, and I feel like I'm being scolded even though I haven't done anything wrong yet.

(2) When I type "5x^2 +", I get "+ is not a valid postfix operator". Most students are going to have no clue what that means. I think you'll need to work on making these messages comprehensible to the average student.

(3) The question mark at the bottom right does not appear to be working properly in the exploration preview tab, especially if you've scrolled down. It shows stuff all the way at the top of the page.

(4) Explorations using the new interaction now have a "learner's answers" section that has rules that say "if the learner's answer ... matches exactly with x=12-5 both...". This is not an OK way to display the shortened form of the rule -- the shortened rule should read "matches exactly with x = 12 - 5 on both sides".

(5) When opening the rules for migrated explorations, the math expression field is empty (although it should contain something like x = 12 - 5). This seems like a serious bug.

Please fix this and any other related issues, then feel free to reassign me and I can run it again. We can't ship the interaction as-is, because the above issues severely impact the usability of what you have built.

vojtechjelinek commented:
Since changes suggested by @seanlip will probably require a lot of refactoring, unassigning code owners for now.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
resolved merge conflicts

iamprayush committed:
All aroung error fixes.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed frontend failures.

iamprayush commented:
> (1) When I type stuff like 5x^2 + ..., a red error message appears and disappears at the bottom while I'm typing. This is very disconcerting, and I feel like I'm being scolded even though I haven't done anything wrong yet.

I have changed this behavior to be less annoying. This is then new behaviour:
![on submit error](https://user-images.githubusercontent.com/35144226/88489419-c3d68b80-cfb1-11ea-8a4d-be280ec7586c.gif)


> (2) When I type "5x^2 +", I get "+ is not a valid postfix operator". Most students are going to have no clue what that means. I think you'll need to work on making these messages comprehensible to the average student.

I have changed the error messages as discussed.

> (3) The question mark at the bottom right does not appear to be working properly in the exploration preview tab, especially if you've scrolled down. It shows stuff all the way at the top of the page.

Fixed this so that it gets displayed properly.

> (4) Explorations using the new interaction now have a "learner's answers" section that has rules that say "if the learner's answer ... matches exactly with x=12-5 both...". This is not an OK way to display the shortened form of the rule -- the shortened rule should read "matches exactly with x = 12 - 5 on both sides".

Changed this so that the human-readable form is displayed. So the rule now reads "matches exactly with x = 12 - 5 on both sides".

> (5) When opening the rules for migrated explorations, the math expression field is empty (although it should contain something like x = 12 - 5). This seems like a serious bug.

Fixed this.

nishantwrp reviewed (approved):
lgtm for codeowner files.

seanlip commented on core/templates/services/guppy-initialization.service.ts:
```diff
@@ -35,16 +37,30 @@ export class GuppyInitializationService {
       // Create a new guppy instance for that div.
       guppyInstance = new Guppy(divId, {});
+
+      // Initialize if with a value for the creator's view.
```
This sentence doesn't scan -- is there a typo or missing word?

iamprayush replied:
It was a typo. I've fixed it now.

seanlip commented on core/templates/services/guppy-initialization.service.ts:
```diff
@@ -35,16 +37,30 @@ export class GuppyInitializationService {
   private onScreenKeyboardShown = false;
 
-  init(guppyDivClassName: string): void {
+  init(guppyDivClassName: string, initialValue = ''): void {
```
Missing typeinfo -- shouldn't a ts checker catch this? /cc @nishantwrp @vojtechjelinek

nishantwrp replied:
Again the type is inferred here due to `initialValue = ''` i.e. string.

seanlip replied:
Ah ok, thanks. What do you want the convention to be in the codebase going forward, though? Should all types be specified, ideally? @nishantwrp

nishantwrp replied:
Hmm, I'm fine with both the possibilities here. But in cases where there is no default value like here, there should be specific types.

seanlip replied:
Ah ok, thanks. Please mention the full desired convention on your types wiki page then, since it will be used as a reference in the future for any type-related issues.  Thanks!

nishantwrp replied:
Done

seanlip commented on core/templates/services/guppy-initialization.service.ts:
```diff
@@ -35,16 +37,30 @@ export class GuppyInitializationService {
+
+      // Initialize if with a value for the creator's view.
+      if (guppyDivClassName === 'guppy-div-creator') {
+        initialValue = initialValue.replace(/abs\(/g, 'absolutevalue(');
```
Are these the only two cases? It should actually be done more declaratively, with a constant mapping and a for loop (and that constant should be used everywhere it needs to be).

iamprayush replied:
Yes, I've tested all the functions that we support and these are the only two cases that need to be handled.
Now that this code is only present in this file, I don't need to add it in a constant, right?

seanlip commented on extensions/objects/templates/numeric-expression-editor.component.ts:
```diff
@@ -41,23 +41,24 @@ angular.module('oppia').component('numericExpressionEditor', {
+        // and 'squareroot(x)' with 'sqrt(x)' for compatibility with nerdamer
+        // and backend validators.
+        ctrl.value = ctrl.value.replace(/absolutevalue\(/g, 'abs(');
+        ctrl.value = ctrl.value.replace(/neg\(/g, '-(');
```
Why is neg handled here but not in the init() case above?

This is why we need a constant -- although now I'm thinking it actually ought to be a service, since you've got stuff after this that is common between all the invocations, too.

iamprayush replied:
This code is no longer needed and I've removed it from all three interactions.

seanlip commented on typings/math-expressions-defs-1.7.0.d.ts:
```diff
@@ -4,6 +4,7 @@ interface Expression {
   derivative: (variable: string) => string;
   equals: (expression: Expression) => boolean;
   toLatex: () => string;
+  toXML: () => string;
```
Why is this needed?

iamprayush replied:
Not needed anymore. I've removed it.

seanlip reviewed (changes requested):
Thanks @iamprayush, took a pass!

seanlip commented:
Also, your CI frontend tests are failing.

seanlip commented:
For (3), the modal is still too high. Try moving it down more so that it is in a position that's similar to our other modals. (If you could use the existing modal infrastructure for that, that would be better too, but it's fine to punt that to a future PR if you like. But the "distance from the top border of the window" issue should be fixed.)

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

seanlip commented:
More issues, apart from the follow-up to point (3) above:

6. The validation job fails with `AssertionError: Validation failed: is_valid_algebraic_expression ({}) for object (4 * 4)`.

7. I tried opening the rule formula box for an existing exploration, deleting all the stuff from it, typing "w2d" (invalid), going to the preview tab without saving the rule, and coming back to the editor tab. Two issues arose:

      - (7a) The old validation behaviour still exists in the place where you enter the rule in "Learner's Answers and Oppia's Responses". I get an error message after almost every keystroke.

      - (7b) I got these errors -- I think you might not be handling the validity check correctly prior to saving:
        ![Screenshot from 2020-07-26 15-08-47](https://user-images.githubusercontent.com/10575562/88490776-d9689880-cf52-11ea-98c8-a340cefc3900.png)

8. Similar to the above, I tried typing in "0asd" for a numeric expression rule formula. It actually saved, which surprised me, and then errored with similar messages to the above. This shouldn't be allowed to be saveable in the first place.

9. Could you please make the placeholder text for the formula box "Type an expression here." or "Type an equation here." or "Type an expression here, using only numbers.", rather than the very generic "Type a formula here."?

ankita240796 reviewed (approved):
LGTM from codeowner's perspective, Thanks @iamprayush!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Handled invalid solutions in audit job.

iamprayush committed:
Minor fix.

iamprayush committed:
Added try except.

iamprayush committed:
temp commit

iamprayush committed:
Fixed migration job failures due to invalid solutions.

iamprayush committed:
temp commit.

iamprayush committed:
temp commit.

iamprayush committed:
Modified audit job,

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

vojtechjelinek reviewed (approved):
LGTM!

iamprayush committed:
More fixes.

nithusha21 commented on assets/constants.ts:
```diff
@@ -5310,6 +5310,21 @@ export = {
+  // Name to human readable form mapping of the position of terms object.
+  "POSITION_OF_TERMS_MAPPING": [{
+    "name": "lhs",
+    "humanReadableName": "on LHS"
```
Should you actually expand LHS in the human readable name?

iamprayush replied:
Done.

nithusha21 commented on core/templates/services/math-interactions.service.spec.ts:
```diff
@@ -124,36 +124,36 @@ describe('MathInteractionsService', () => {
       '3.4.5 + 45/a')).toBeFalse();
     expect(mathInteractionsService.getWarningText()).toBe(
-      'Invalid integer.');
+      'Your answer contains an invalid integer: 3.4.5');
```
invalid term you mean? 3.4.5 is not an integer

iamprayush replied:
Yup, makes sense. Done.

nithusha21 commented on core/templates/services/math-interactions.service.spec.ts:
```diff
@@ -186,7 +186,7 @@ describe('MathInteractionsService', () => {
     expect(mathInteractionsService.validateEquation('')).toBeFalse();
     expect(mathInteractionsService.getWarningText()).toBe(
-      'Please enter a non-empty answer.');
+      'Your answer seems to be empty.');
```
"seems to be empty" is probably not right. Maybe you can say "Please enter an answer before submitting"

iamprayush replied:
Done.

nithusha21 commented on core/templates/services/math-interactions.service.spec.ts:
```diff
@@ -247,7 +247,7 @@ describe('MathInteractionsService', () => {
       '3.4.5 = 45/a')).toBeFalse();
     expect(mathInteractionsService.getWarningText()).toBe(
-      'Invalid integer.');
+      'Your answer contains an invalid integer: 3.4.5');
```
ditto, reg integer

iamprayush replied:
Done.

nithusha21 commented on extensions/interactions/MathEquationInput/directives/oppia-interactive-math-equation-input.component.spec.ts:
```diff
@@ -129,7 +129,7 @@ describe('MathEquationInputInteractive', function() {
     ctrl.value = '';
     expect(ctrl.isCurrentAnswerValid()).toBeFalse();
-    expect(ctrl.warningText).toBe('Please enter a non-empty answer.');
+    expect(ctrl.warningText).toBe('Your answer seems to be empty.');
```
Ditto, reg empty

iamprayush replied:
Done.

nithusha21 reviewed (approved):
LGTM, just some minor nits!

iamprayush committed:
Fixed frontend tests.

iamprayush committed:
Fixed linter errors.

iamprayush commented:
I'd compiled a list of issues raised by @seanlip and @aks681 to keep track. I've solved them all:

### Problems:
- [x] Validaiton job failure.

Fixed.
- [x] Modal is too high up on the page

Centered it now and also reduced the width to fit-content.
- [x] Fix error messages showing up annoyingly in creator's view as well

The save button gets disabled in real-time but the warnings show only when the editor is not in focus.
- [x] Typing a rule input and going to the preview tab without saving gives errors

Fixed. Stored the current value in an intermediate variable instead of directly updating the ctrl.value since leaving the page auto saves ctrl.value even if its invalid which was causing the error.
- [x] Customize the placeholder text based on the type of intreaction

Done.
- [x] Typing "0asd" in numeric interaction allows saving and then gives error.

This was happening because I'm using nerdamer to fetch variables from the expression but "0asd" gets evaluated as "0" since it simplifies numebrs wherever it can, so I've now used a manual check instead and it works as expected.
- [x] With bracketed terms, ^2 and ^3  doesn't seem to work

Fixed.
- [x] Add = sign in OSK

Added.
- [x] Editing saved rules for math equations didn't work.

Fixed.

iamprayush commented:
@U8NWXD PTAL as code-owner, there are some minor changes in the e2e test files.

seanlip reviewed (approved):
LGTM. Job is successful!

seanlip commented:
Any idea what's going on with the missing CircleCI tests?

U8NWXD reviewed (approved):
@iamprayush the e2e test changes look great! Since there aren't any logic or flow changes, I think we can skip the flakiness evaluation

seanlip commented:
@iamprayush please resolve the merge conflicts!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Removed draft-changes one-off job,

seanlip reviewed (changes requested):
Found an issue: written translation ids are not being handled properly when feedback is deleted (when handling this case, please make sure the changelist follows the exact set of operations that would be done in the editor).

EDIT: sorry, scratch that last point. I forgot that, in migrations, the change cmds are handled differently.

iamprayush committed:
Handled deleted feedbacks.

iamprayush committed:
Replaced cdot with times

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into upgrade-old-explorations

iamprayush committed:
Fixed linter errors.

iamprayush commented:
> Found an issue: written translation ids are not being handled properly when feedback is deleted (when handling this case, please make sure the changelist follows the exact set of operations that would be done in the editor).
> 
> EDIT: sorry, scratch that last point. I forgot that, in migrations, the change cmds are handled differently.

I've fixed it now.

iamprayush committed:
Fixed frontend failures

iamprayush committed:
Minor fix.

seanlip reviewed (approved):
This PR is approved for merge. Please merge ASAP. Thanks!
